test/integration/Utils.kt

```diff
@@ -15,20 +15,181 @@
+    return absolute.toPath()
+}
+
+/// Wait `timeMillis` milliseconds, then wait for all recompositions to finish.
+suspend fun wait(composeRule: ComposeContentTestRule, timeMillis: Int) {
+    delay(timeMillis.toLong())
+    composeRule.waitForIdle()
+}
+
+suspend fun connectToTypeDB(composeRule: ComposeContentTestRule, address: String) {
+    // This opens a dialog box (which we can't see through) so we assert that buttons with that text can be
+    // clicked.
+    composeRule.onAllNodesWithText(Label.CONNECT_TO_TYPEDB).assertAll(hasClickAction())
+
+    StudioState.client.tryConnectToTypeDB(address) {}
+    // Resolving localhost can take up to 5 seconds on macOS
+    wait(composeRule, 5_000)
+    assertTrue(StudioState.client.isConnected)
+
+    composeRule.onNodeWithText(address).assertExists()
+}
+
+suspend fun createDatabase(composeRule: ComposeContentTestRule, name: String) {
+    composeRule.onAllNodesWithText(Label.SELECT_DATABASE).assertAll(hasClickAction())
+
+    StudioState.client.tryDeleteDatabase(name)
+    wait(composeRule, 500)
+
+    StudioState.client.tryCreateDatabase(name) {}
+    wait(composeRule,500)
+}
+
+suspend fun writeSchemaInteractively(composeRule: ComposeContentTestRule, database: String, fileName: String) {
+    composeRule.onNodeWithText(PLUS_ICON_STRING).performClick()
+    wait(composeRule, 500)
+
+    StudioState.client.session.tryOpen(database, TypeDBSession.Type.SCHEMA)
+    wait(composeRule, 500)
+
+    StudioState.client.tryUpdateTransactionType(TypeDBTransaction.Type.WRITE)
+    wait(composeRule, 500)
+
+    composeRule.onNodeWithText("schema").performClick()
+    composeRule.onNodeWithText("write").performClick()
+
+    StudioState.project.current!!.directory.entries.find { it.name == fileName }!!.asFile().tryOpen()
+
+    composeRule.onNodeWithText(PLAY_ICON_STRING).performClick()
+    wait(composeRule, 500)
+
+    composeRule.onNodeWithText(CHECK_ICON_STRING).performClick()
+    wait(composeRule, 500)
+}
+
+suspend fun writeDataInteractively(composeRule: ComposeContentTestRule, database: String, fileName: String) {
+    StudioState.client.session.tryOpen(database, TypeDBSession.Type.DATA)
+
+    wait(composeRule, 500)
+
+    composeRule.onNodeWithText("data").performClick()
+    composeRule.onNodeWithText("write").performClick()
+
+    StudioState.project.current!!.directory.entries.find { it.name == fileName }!!.asFile().tryOpen()
+
+    composeRule.onNodeWithText(PLAY_ICON_STRING).performClick()
+    wait(composeRule, 500)
+
+    composeRule.onNodeWithText(CHECK_ICON_STRING).performClick()
+    wait(composeRule, 500)
+}
+
+suspend fun verifyDataWrite(composeRule: ComposeContentTestRule, dbName: String, queryFilePath: String) {
+    val queryString = fileNameToString(queryFilePath)
+
+    composeRule.onNodeWithText("infer").performClick()
+    composeRule.waitForIdle()
+    composeRule.onNodeWithText("read").performClick()
+    composeRule.waitForIdle()
+    delay(1_000)
```

Comment:
Just noticed this - we can use `wait(composeRule, 1_000)` right? (And should because doing the delay before waiting for recomposition is the better flow, if I remember right)

Reply:
Absolutely. This has been done.